packages/SwingSet/docs/run-policy.md

```diff
@@ -0,0 +1,132 @@
+
+But the more sophisticated approach is to call `controller.run(policy)`, with a "Run Policy" object that knows when the block should end. The policy object will get detailed information about each delivery, including the metering results (how many low-level JS engine operations were performed), and can use this to guide the block size.
+
+## Cranks
+
+The kernel maintains two queues. The highest priority queue contains "GC Actions", which are message to vats that indicate an object has been garbage collected and can now be freed. These are provoked by reference-counting operations that occur as a side-effect of GC syscalls, as well as vat termination events. Each GC Action counts as a "crank", and is reported to the policy object. The policy may end a block while there is still GC Action work to do, in which case the kernel will pick it back up again when the next block begins.
+
+If the GC Action queue is entirely empty, the kernel will look for regular work to do. This consists of the following event types:
+
+* message deliveries to vats (provoked by `syscall.send`, delivered as `dispatch.deliver`)
+* promise resolution notifications (provoked by `syscall.resolve`, delivered as `dispatch.notify`)
+* vat creation
+
+Each message delivery and resolution notification causes (at most) one vat to execute one "crank" (it might not execute any crank, e.g. when a message is delivered to an unresolved promise, it just gets added to the promise's queue). This crank gives the vat some amount of time to process the delivery, during which it may invoke any number of syscalls. The crank might cause the vat to be terminated, either because of an error, or because it took too much CPU and exceeded its Meter's allowance. Each crank yields a "delivery results object", which indicates the success
+
+When run in a suitable vat worker (`managerType: 'xs-worker'`), the delivery results include the number of "computrons" consumed, as counted by the JS engine. Computrons are vaguely correlated to CPU cycles (despite being much larger) and thus some correspondence to wallclock time. A Run Policy which wishes to limit wallclock time in a consensus-base manner should pay attention to the cumulative computron count, and end the block after some experimentally-determined limit.
+
+Vat creation also gives a single vat (the brand new one) time to run the top-level forms of its source bundle, as well as the invocation of its `buildRootObject()` method. This typically takes considerably longer than the subsequent messages, and is not currently metered.
+
+## Run Policy
+
+The kernel will invoke the following methods on the policy object (so all must exist, even if they're empty):
+
+* `policy.vatCreated()`
+* `policy.crankComplete(computrons)`
+* `policy.crankFailed()`
+
+All methods should return `true` if the kernel should keep running, or `false` if it should stop.
+
```

Comment:
It's a little weird that if the return from `crankFailed` is true the vat should keep going.

Reply:
Does it help to know that it's the *kernel* which should keep running, rather than the vat? Yeah, at that point the vat is dead. We don't get a lot of data about how long it took the vat to fail: if the vat exceeds its per-crank limit, the process terminates, and all we get is a unix exit status.

Reply:
There's an ambiguity in the interpretation of the name -- it could mean "what do I do if if the crank failed?" or it could me "did the crank fail?"